README.md
Cooperative
==============

Create concurrent code without thinking in callbacks.



CHANGELOG.rst

## 0.1.10 (April 10, 2015)

Features:

  - Refactor an internal function.

Bugfixes:

  - Use PipSession as a context manager so it automatically closes a session.


cooperative/tests/test_version.py
import unittest


class TestMeta(unittest.TestCase):
    def test_version_defined(self):
        """
        Ensure __version__ is defined in the stream_tap module.

        Ensure __version_info__ is defined in the stream_tap module.

        """
        from stream_tap import __version__
        from stream_tap import __version_info__
        self.assertTrue(__version__)
        self.assertTrue(__version_info__)


cooperative/tests/test_cooperative.py
# _*_ coding: utf-8 _*_
from collections import deque
from itertools import chain

from twisted.internet import defer
from twisted.internet.defer import inlineCallbacks
from twisted.internet.task import Clock
from twisted.internet.task import Cooperator
from twisted.internet.task import LoopingCall
from twisted.python import log
from twisted.trial import unittest

from stream_tap import Bucket

from cooperative import accumulation_handler
from cooperative import accumulate
from cooperative import batch_accumulate


class TestHandler(unittest.TestCase):
    def test_accumulation_handler(self):
        """
        Ensure the return value of accumulation_handler
        is the contents of a Bucket instance with it's contents
        drained.

        :return:
        """
        spigot = Bucket(unicode)
        spigot(1)
        spigot(2)

        result = accumulation_handler(None, spigot)
        self.assertEqual(result, deque([u"1", u"2"]))
        self.assertEqual(spigot.contents(), deque([]))


def i_get_tenth_11(value):
    """
    Yield the tenth and eleventh item of value.

    :param value:
    :return:
    """
    yield value[10]
    yield value[11]


@inlineCallbacks
def run_some_with_error(value):
    """
    Cooperatively iterator over two iterators consecutively, but
     the second one will always raise an IndexError, which is caught,
     logged and a message is returned.

    :return:
    """
    #  first deferred
    result = yield accumulate(i_get_tenth_11(value))

    try:
        result2 = yield accumulate(i_get_tenth_11(result))
        defer.returnValue(result2)
    except IndexError, e:
        log.err(e)
        defer.returnValue("Couldn't get second result, "
                          "but the first result is {}".format(result))


@inlineCallbacks
def run_some_without_error(value):
    """
    Cooperatively iterator over two iterators consecutively and the
    result of the final one is returned.

    :param value: Any sequence.
    :return:
    """
    #  first deferred
    result = yield accumulate(i_get_tenth_11(range(110, 150)))

    log.msg("accumulated {}".format(result))
    result = yield accumulate(i_get_tenth_11(value))
    defer.returnValue(result)


class Doer(object):
    count = 0

    def __init__(self, own_reactor, own_cooperator):
        self.reactor = own_reactor
        self.cooperator = own_cooperator
        self.previous = deque([], maxlen=50)
        self.count = 0

    @inlineCallbacks
    def run(self):
        """
        Cooperatively iterator over two iterators consecutively and the
        result of the final one is returned.

        :return:
        """
        #  first deferred
        self.count += 1
        result = yield accumulate(i_get_tenth_11(range(110, 150)), self.cooperator)

        log.msg("accumulated {} at {}".format(result, self.count))
        result = yield batch_accumulate(2, range(self.count), self.cooperator)
        log.msg("batch accumulated {} at {}".format(result, self.count))
        defer.returnValue(result)


class TestOwnCooperator(unittest.TestCase):
    def setUp(self):
        """
        Create a reactor and Cooperator that can be controlled.

        Instantiate a Doer with the reactor and cooperator.

        Create a Looping Call and set it's clock to the reactor.

        :return:
        """
        self.reactor = Clock()
        self.cooperator = Cooperator()
        self.doer = Doer(self.reactor, self.cooperator)
        self.loop = LoopingCall(self.doer.run)
        self.loop.clock = self.reactor

    def test_control_coop(self):
        """
        Ensure control of own cooperator.

        :return:
        """
        def tick():
            self.reactor.advance(1)
            self.cooperator._tick()

        self.loop.start(1.0, now=True)
        tick()
        self.assertEqual(1, self.doer.count)
        tick()
        self.assertEqual(2, self.doer.count)
        tick()
        self.assertEqual(3, self.doer.count)

    def tearDown(self):
        self.loop.stop()


class TestAccumulate(unittest.TestCase):
    @inlineCallbacks
    def test_accumulate(self):
        """
        Ensure that within an inline callback function,
        a accumulate wrapped generator
        yields the result of the output of the generator.

        :return:
        """
        result = yield accumulate(i_get_tenth_11(range(110, 150)))
        self.assertEqual(result, deque([120, 121]))


    @inlineCallbacks
    def test_failure(self):
        """
        Ensure that within an inline callback function,
        a accumulate based function
        yields the result if it's cooperative generator.

        Since and_the_winner_is is designed to always
        log and error, Ensure one IndexError is logged.

        :return:
        """
        result = yield run_some_with_error(range(15))
        self.assertEqual(result, "Couldn't get second result, "
                                 "but the first result is deque([10, 11])")

        # failure expected
        index_errors = self.flushLoggedErrors(IndexError)
        self.assertEqual(len(index_errors), 1)

    @inlineCallbacks
    def test_multi_winner(self):
        """
        Ensure multiple inline callback functions will run cooperatively.

        :return:
        """
        d1 = run_some_without_error(range(15))
        d2 = run_some_without_error(range(15, 100))
        result = yield defer.gatherResults([d1, d2])

        self.assertEqual(result, [deque([10, 11]), deque([25, 26])])

    @inlineCallbacks
    def test_trice_winner(self):
        """
        Ensure multiple inline callback functions will run cooperatively.

        :return:
        """
        d1 = run_some_without_error(range(15))
        d2 = run_some_without_error(range(15, 100))
        d3 = accumulate(i_get_tenth_11(range(4, 50)))
        result = yield defer.gatherResults([d1, d2, d3])

        self.assertEqual(result, [deque([10, 11]),
                                  deque([25, 26]),
                                  deque([14, 15])])

    @inlineCallbacks
    def test_multi_winner_chain(self):
        """
        Ensure multiple inline callback functions will run cooperatively.

        Ensure the result of gatherResults can be chained together
        in order.

        :return:
        """
        called = []

        def watcher(value):
            """
            A pass through generator for i_get_tenth_11 that
            captures the value in a list, which can show the
            order of the generator iteration.

            :param value:
            :return:
            """
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item

        result = yield defer.gatherResults([
            accumulate(watcher(range(0, 15))),
            accumulate(watcher(range(15, 100))),
            accumulate(watcher(range(98, 200))),
            accumulate(watcher(range(145, 189)))
        ])

        final_result = list(chain.from_iterable(result))

        self.assertEqual(final_result, [10, 11, 25, 26, 108, 109, 155, 156])
        self.assertEqual(set(called), set(final_result))
        self.assertNotEqual(called, final_result)
        #  The iteration is shown to alternate between generators passed
        #   to cooperate.
        self.assertEqual(called, [10, 25, 108, 155, 11, 26, 109, 156])

    @inlineCallbacks
    def test_multi_deux_chain(self):
        """
        Ensure multiple inline callback functions will run cooperatively.

        Ensure the result of gatherResults can be chained together
        in order.

        Ensure cooperatively run generators will complete
        no matter the length.

        Ensure the longest one will continue to iterate after the
        others run out of iterations.

        :return:
        """
        called = []

        def watcher(value):
            """
            A pass through generator for i_get_tenth_11 that
            captures the value in a list, which can show the
            order of the generator iteration.

            :param value:
            :return:
            """
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item

        def deux_watcher(value):
            """
            A pass through generator for i_get_tenth_11 that
            captures the value in a list, which can show the
            order of the generator iteration.

            :param value:
            :return:
            """
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item

        result = yield defer.gatherResults([
            accumulate(watcher(range(0, 15))),
            accumulate(watcher(range(15, 100))),
            accumulate(deux_watcher(range(1098, 10200))),
            accumulate(watcher(range(145, 189)))
        ])

        final_result = list(chain.from_iterable(result))

        self.assertEqual(final_result,
                         [10, 11, 25, 26, 1108, 1109, 1108, 1109, 155, 156])
        self.assertEqual(set(called), set(final_result))
        self.assertNotEqual(called, final_result)
        #  The iteration is shown to alternate between generators passed
        #   to cooperate.
        self.assertEqual(called,
                         [10, 25, 1108, 155, 11, 26, 1109, 156, 1108, 1109])

    @inlineCallbacks
    def test_multi_deux_batched(self):
        """
        Ensure multiple inline callback functions will run cooperatively.

        Ensure the result of gatherResults can be chained together
        in order.

        Ensure cooperatively run generators will complete
        no matter the length.

        Ensure the longest one will continue to iterate after the
        others run out of iterations.

        Ensure those called with batch_accumulate will
        iterate over the generator in batches the size of max_size.

        :return:
        """
        called = []

        def watcher(value):
            """
            A pass through generator for i_get_tenth_11 that
            captures the value in a list, which can show the
            order of the generator iteration.

            :param value:
            :return:
            """
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item

        def deux_watcher(value):
            """
            A pass through generator for i_get_tenth_11 that
            captures the value in a list, which can show the
            order of the generator iteration.

            :param value:
            :return:
            """
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item
            for item in i_get_tenth_11(value):
                called.append(item)
                yield item

        result = yield defer.gatherResults([
            accumulate(watcher(range(0, 15))),
            accumulate(watcher(range(15, 100))),
            batch_accumulate(3, deux_watcher(range(1098, 10200))),
            accumulate(watcher(range(145, 189)))
        ])

        final_result = list(chain.from_iterable(result))

        self.assertEqual(final_result,
                         [10, 11, 25, 26, 1108, 1109, 1108, 1109, 155, 156])
        self.assertEqual(set(called), set(final_result))
        self.assertNotEqual(called, final_result)
        #  The iteration is shown to alternate between generators passed
        #   to cooperate, except the batched one does three at a time
        #   when it is it's turn.
        self.assertEqual(called,
                         [10, 25, 1108, 1109, 1108, 155, 11, 26, 1109, 156])


examples/blocking.py
#!/usr/bin/env python
# _*_ coding: utf-8 _*_
from operator import add

import add_package_path

import sys
from twisted.internet import defer
from twisted.internet.task import react
from twisted.python import log

from cooperative import batch_accumulate


def expensive(number):
    log.msg("starting {}".format(number))
    for value in range(100000):
        if 25000 == value:
            log.msg("1/4 for {}".format(number))
        if 50000 == value:
            log.msg("1/2 for {}".format(number))
        if 75000 == value:
            log.msg("3/4 for {}".format(number))
        yield number * value / 3.0


@defer.inlineCallbacks
def do_some_expensive_things(number):
    """
    Perform one expensive computation cooperatively with any
     other iterator passed into twisted's cooperate, then
     use it's result to pass into the second computation.

    :param number:
    :return:
    """
    result = yield batch_accumulate(1000, expensive(number))
    total = reduce(add, result, 0)
    log.msg("first for {}: {}".format(number, total))

    result = yield batch_accumulate(1000, expensive(int(total/1e9)))
    total = reduce(add, result, 0)
    log.msg("second for {}: {}".format(number, total))
    defer.returnValue(total)


def main(reactor):

    d5 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(0.3, d5.callback, "########## simulated request 1 ############")

    d6 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(0.5, d6.callback, "########## sim request 2 ############")

    d7 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(0.7, d7.callback, "########## simulated request 3 ############")

    # simulate an external event triggering an expensive computation while
    # other expensive computations are happening.
    d8 = defer.Deferred()
    d8.addCallback(do_some_expensive_things)
    reactor.callLater(0.1, d8.callback, 201)

    numbers = [54.0, 42, 10, 34]
    for number in numbers:
        result = list(expensive(number))
        log.msg("first for {}: {}".format(number, reduce(add, result, 0)))
        result = list(expensive(number))
        log.msg("second for {}: {}".format(number, reduce(add, result, 0)))

    return defer.gatherResults([d5, d6, d7, d8]).addCallback(log.msg)

if __name__ == "__main__":
    log.startLogging(sys.stdout)
    react(main, [])


cooperative/__init__.py
# _*_ coding: utf-8 _*_
# -*- test-case-name: cooperative.tests.test_cooperative -*-
from collections import deque

from cooperative import _meta

__version__ = _meta.version
__version_info__ = _meta.version_info


from twisted.internet.task import cooperate

from stream_tap import stream_tap
from iter_karld_tools import i_batch


class ValueBucket(object):
    """
    Produces a callable that accumulates all non-None values
    it is called with in order.

    The contents may be accessed or collected and drained,
    to make room for new content.
    """
    def __init__(self):
        self._contents = deque()

    def __call__(self, value):
        if value is not None:
            self._contents.append(value)

    def contents(self):
        """
        :returns: contents
        """
        return self._contents

    def drain_contents(self):
        """
        Starts a new collection to accumulate future contents
        and returns all of existing contents.
        """
        existing_contents = self._contents
        self._contents = deque()
        return existing_contents


def accumulation_handler(stopped_generator, spigot):
    """
    Drain the contents of the bucket from the spigot.

    :param stopped_generator: Generator which as stopped
    :param spigot: a Bucket.
    :return: The contents of the bucket.
    """
    return spigot.drain_contents()


def accumulate(a_generator, cooperator=None):
    """
    Start a Deferred whose callBack arg is a deque of the accumulation
    of the values yielded from a_generator.

    :param a_generator: An iterator which yields some not None values.
    :return: A Deferred to which the next callback will be called with
     the yielded contents of the generator function.
    """
    if cooperator:
        own_cooperate = cooperator.cooperate
    else:
        own_cooperate = cooperate

    spigot = ValueBucket()
    items = stream_tap((spigot,), a_generator)
    d = own_cooperate(items).whenDone()
    d.addCallback(accumulation_handler, spigot)
    return d


def batch_accumulate(max_batch_size, a_generator, cooperator=None):
    """
    Start a Deferred whose callBack arg is a deque of the accumulation
    of the values yielded from a_generator which is iterated over
    in batches the size of max_batch_size.

    It should be more efficient to iterate over the generator in
     batches and still provide enough speed for non-blocking execution.

    :param max_batch_size: The number of iterations of the generator
     to consume at a time.
    :param a_generator: An iterator which yields some not None values.
    :return: A Deferred to which the next callback will be called with
     the yielded contents of the generator function.
    """
    if cooperator:
        own_cooperate = cooperator.cooperate
    else:
        own_cooperate = cooperate

    spigot = ValueBucket()
    items = stream_tap((spigot,), a_generator)

    d = own_cooperate(i_batch(max_batch_size, items)).whenDone()
    d.addCallback(accumulation_handler, spigot)
    return d


examples/non_blocking.py
#!/usr/bin/env python
# _*_ coding: utf-8 _*_
from operator import add

import add_package_path

import sys
from twisted.internet import defer
from twisted.internet.task import react
from twisted.python import log

from cooperative import batch_accumulate


def expensive(number):
    log.msg("starting {}".format(number))
    for value in range(100000):
        if 25000 == value:
            log.msg("1/4 for {}".format(number))
        if 50000 == value:
            log.msg("1/2 for {}".format(number))
        if 75000 == value:
            log.msg("3/4 for {}".format(number))
        yield number * value / 3.0


def expensive2(number):
    log.msg("starting {}".format(number))
    total = 0
    for value in range(100000):
        if 25000 == value:
            log.msg("1/4 for {}".format(number))
        if 50000 == value:
            log.msg("1/2 for {}".format(number))
        if 75000 == value:
            log.msg("3/4 for {}".format(number))
        total += number * value / 3.0
        yield
    yield total


@defer.inlineCallbacks
def do_some_expensive_things(number):
    """
    Perform one expensive computation cooperatively with any
     other iterator passed into twisted's cooperate, then
     use it's result to pass into the second computation.

    :param number:
    :return:
    """
    result = yield batch_accumulate(1000, expensive(number))
    total = reduce(add, result, 0)
    log.msg("first for {}: {}".format(number, total))

    result = yield batch_accumulate(1000, expensive2(int(total/1e9)))
    total = reduce(add, result, 0)
    log.msg("second for {}: {}".format(number, total))
    defer.returnValue(total)


@defer.inlineCallbacks
def do_less_expensive_things(number):
    """
    Perform one expensive computation cooperatively with any
     other iterator passed into twisted's cooperate, then
     use it's result to pass into the second computation.

    :param number:
    :return:
    """
    result = yield batch_accumulate(1000, expensive(number))
    total = reduce(add, result, 0)
    log.msg("only for {}: {}".format(number, total))
    defer.returnValue(total)


def main(reactor):
    d1 = do_some_expensive_things(54.0)
    d2 = do_some_expensive_things(42)
    d3 = do_some_expensive_things(10)
    d4 = do_some_expensive_things(34)

    # Enqueue events to simulate handling external events
    d5 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(0.3, d5.callback, "########## simulated request 1 ############")

    d6 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(0.5, d6.callback, "########## sim request 2 ############")

    d7 = defer.Deferred().addCallback(log.msg)
    reactor.callLater(1.0, d7.callback, "########## simulated request 3 ############")

    # simulate an external event triggering an expensive computation while
    # other expensive computations are happening.
    d8 = defer.Deferred()
    d8.addCallback(do_less_expensive_things)
    reactor.callLater(0.3, d8.callback, 20001)

    return defer.gatherResults([d1, d2, d3, d4, d5, d6, d7, d8]).addCallback(log.msg)

if __name__ == "__main__":
    log.startLogging(sys.stdout)
    react(main, [])
